libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,161 +1681,198 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+					(
+						(
+							resultArrayType->isPointer() ||
+							(argArrayType->isByteArray() && resultArrayType->isByteArray())
+						) &&
+						resultArrayType->location() == DataLocation::Storage
+					),
+					"Invalid explicit conversion to storage type."
+				);
 			}
-			if (resultType->category() == Type::Category::Address)
+		}
+		else
+		{
+			if (
+				resultType->category() == Type::Category::Contract &&
+				argType->category() == Type::Category::Address
+			)
 			{
-				bool payable = argType->isExplicitlyConvertibleTo(AddressType::addressPayable());
-				resultType = make_shared<AddressType>(payable ? StateMutability::Payable : StateMutability::NonPayable);
+				solAssert(dynamic_cast<ContractType const*>(resultType.get())->isPayable(), "");
+				solAssert(
+					dynamic_cast<AddressType const*>(argType.get())->
+						stateMutability() < StateMutability::Payable,
+					""
+				);
+				SecondarySourceLocation ssl;
+				if (
+					auto const* identifier =
+						dynamic_cast<Identifier const*>(arguments.front().get())
+				)
+					if (
+						auto const* variableDeclaration = dynamic_cast<VariableDeclaration const*>(
+							identifier->annotation().referencedDeclaration
+						)
+					)
+						ssl.append(
+							"Did you mean to declare this variable as \"address payable\"?",
+							variableDeclaration->location()
+						);
+				m_errorReporter.typeError(
+					_functionCall.location(), ssl,
+					"Explicit type conversion not allowed from non-payable \"address\" to \"" +
+					resultType->toString() +
+					"\", which has a payable fallback function."
+				);
 			}
+			else
+				m_errorReporter.typeError(
+					_functionCall.location(),
+					"Explicit type conversion not allowed from \"" +
+					argType->toString() +
+					"\" to \"" +
+					resultType->toString() +
+					"\"."
+				);
+		}
+		if (resultType->category() == Type::Category::Address)
+		{
+			bool const payable = argType->isExplicitlyConvertibleTo(AddressType::addressPayable());
+			resultType = make_shared<AddressType>(
+				payable ? StateMutability::Payable : StateMutability::NonPayable
+			);
 		}
-		_functionCall.annotation().type = resultType;
-		_functionCall.annotation().isPure = isPure;
-
-		return false;
 	}
+	return resultType;
+}
 
-	// Actual function call or struct constructor call.
+TypePointer TypeChecker::typeCheckFunctionLikeAndRetrieveReturnType(
+	FunctionCall const& _functionCall,
+	FunctionTypePointer _functionType
+)
+{
+	if (!_functionType)
+	{
+		m_errorReporter.typeError(_functionCall.location(), "Type is not callable");
+		return make_shared<TupleType>();
+	}
 
-	FunctionTypePointer functionType;
+	TypePointers returnTypes;
 
-	/// For error message: Struct members that were removed during conversion to memory.
-	set<string> membersRemovedForStructConstructor;
-	if (_functionCall.annotation().kind == FunctionCallKind::StructConstructorCall)
+	if (_functionType->kind() == FunctionType::Kind::ABIDecode)
```

Comment:
I think it would be better to pull this out into the calling function

Reply:
I've moved the ABIDecode case handling up to visit().